### DISCUSSION TITLE
[HttpToSocksProxy] How to redirect everything to an upstream SOCKS5 proxy

### DISCUSSION THREAD

**Discussion by 4n1qz5skwv** [abhinavsingh/proxy.py#948] [Q&A] [answered]
I don't the think the  title actually describe it all or correctly.

## Looking For

Say, User set `127.0.0.1:12345` as `proxy` in Firefox. And if the code is running, all the traffic will actually flow through a `sock5` proxy `(socks5://123.456.789.010:25456)`.

## Main Purpose

So my main purpose is to let user use the socks5 proxy but without actually exposing it. I don't know what this method call. So i can't properly Google it. Is this possible using this library?

Please help. Thank you.

**Comment by abhinavsingh** [comment]
I guess you are looking for is a `HttpToSocksProxyPoolPlugin` i.e. something which will provide following tunnel:

`Firefox <--- http proxy -----> 127.0.0.1:12345  <---- socks5 ----> 123.456.789.010:25456`

This is possible but haven't been tried.

Here are a few suggestions if you want to hack around:

1. Take inspiration from [ProxyPoolPlugin](https://github.com/abhinavsingh/proxy.py/blob/develop/proxy/plugin/proxy_pool.py).  This is EXACTLY what you need, but instead of upstream HTTP proxies you want to use SOCKS5 proxies, hence the name `HttpToSocksProxyPoolPlugin` :). You can have one or multiple upstream SOCKS5 proxies, hence `pool` used in the name.
2. For `HttpToSocksProxyPoolPlugin` to work, you will need to translate HTTP proxy packets (client) to SOCKS5 packet (server) and vice-versa.  This is something `ProxyPoolPlugin` already do, but it needs to handle only `HTTP` packets.
3. To carry out SOCKS packet parsing/building, you will need an external library.

So:

1. Copy `ProxyPoolPlugin` as `HttpToSocksProxyPoolPlugin`
2. Update code to send outgoing `HTTP` proxy packets as corresponding `SOCKS5` packets to upstream
3. Update code to parse incoming upstream packets as `SOCKS5` packets and send to clients as HTTP packets

Most of the task should be straightforward except for dealing with SOCKS5 specific authentication and other workflows.  But once connection has been established, you can pretty much tunnel packets as is.

Let me know about your thoughts.  Thank you.

**Reply by 4n1qz5skwv** [reply]
Thank you for your detailed reply. Unfortunately, i know nothing about this library, and when i heard that it **maybe** possible. i dropped it. No way i could have solve it. Also by more research i found this "https://github.com/qwj/python-proxy/issues/112". exactly what i wanted.

**Comment by 4n1qz5skwv** [comment, marked as answer]
Answer: Not directly but may be. Found an alternative.

**Reply by abhinavsingh** [reply]
Thanks you for the update.  I think your use case is certainly possible with `proxy.py`, but there is no baked in solution for such protocol transformation.  This can be a good feature to work upon in the future.

I checked `python-proxy` and it already has baked in `SOCKS` protocol support.  Hence, it's a no-brainer that they can support such use case out of box.  IIUC, you wanted incoming HTTP traffic to proxy to upstream SOCKS proxy.  This will require more effort than just queuing packets back and forth between the clients.  More so because `proxy.py` has no baked in `SOCKS` protocol support.

FWIW, as a matter of chance, today I sent out :)
- https://github.com/abhinavsingh/proxy.py/pull/1047

But, if an existing solution already works for you, I'll recommend just go for it.
